### ISSUE TITLE
function

### ISSUE THREAD

**Issue by rhakb** [souzomain/Shaco#1]
hi, bro
Can you obtain command output through pipelines? This can avoid blocking long running commands

**Comment by souzomain** [comment]
yeah, of course, but pipelines still going block, an alternative is creating threads for commands.

Or did I get it wrong? show me your idea bro

**Comment by rhakb** [comment]
you can create sub processes and then read then through non blocking pipes,which is also how havoc's windows implementation works

**Comment by rhakb** [comment]
sample code

`#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>

#define BUFFER_SIZE 1024

int main() {
    int fd[2];
    char buffer[BUFFER_SIZE];
    int nbytes;

    if (pipe(fd) < 0) {
        perror("pipe");
        exit(EXIT_FAILURE);
    }

    int flags = fcntl(fd[0], F_GETFL, 0);
    fcntl(fd[0], F_SETFL, flags | O_NONBLOCK);  // Set the reader to non blocking

    pid_t pid = fork();
    if (pid < 0) {
        perror("fork");
        exit(EXIT_FAILURE);
    } else if (pid == 0) {  // subprocess
        close(fd[0]);  // Close Readers

        dup2(fd[1], STDOUT_FILENO);  // Redirect standard output to pipeline writer

        execlp("nmap", "nmap", "-sP", "192.168.0.0/24", NULL);  // Perform nmap scan

        perror("execlp");  // Execlp failed
        exit(EXIT_FAILURE);
    } else {  // parent process
        close(fd[1]);  // Close Write End

        while (1) {
            nbytes = read(fd[0], buffer, BUFFER_SIZE);
            if (nbytes < 0) {  // Read error
                perror("read");
            } else if (nbytes == 0) {  // The pipeline has been closed
                printf("End of file\n");
                break;
            } else {  // Data available
                buffer[nbytes] = '\0';
                printf("Received: %s", buffer);
            }

            printf("Do other things\n");
            sleep(1);
        }

        wait(NULL);  // Waiting for child processes to exit
    }

    return 0;
}
`

**Comment by rhakb** [comment]
You can check and read the pipeline once every time you upload data from other operations, and upload it instead of waiting for the command to complete to avoid affecting other operations

**Comment by souzomain** [comment]
Interesting, but the demon of havoc call the function without create new thread or process. 

![image](https://user-images.githubusercontent.com/92044641/236684255-2b9289a2-b5dd-4ccb-80d4-979939f88bc2.png)

I can implement directly into the function when I call or create a new DEMON-like command called Job.
What do you think?

**Comment by souzomain** [comment]
Let's implement this.

but for extented commands, demon of havoc uses job command, I can create this.

do you think best implement this for all commands executed or create "job" command to this?

**Comment by rhakb** [comment]
Sorry, I have removed the fork function and improved the program. Here is the code

```
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>

#define BUFFER_SIZE 1024

int main(int argc, char *argv[]) {
    char buffer[BUFFER_SIZE];
    FILE *nmap_pipe;
    int fd;
    fd_set set;
    struct timeval timeout;

    // Execute the nmap command and open the pipeline
    nmap_pipe = popen("nmap -sP -O 127.0.0.1", "r");
    if (!nmap_pipe) {
        fprintf(stderr, "Failed to open pipe: %s\n", strerror(errno));
        exit(EXIT_FAILURE);
    }

    // Set the pipeline to non blocking mode
    fd = fileno(nmap_pipe);
    fcntl(fd, F_SETFL, O_NONBLOCK);

    //Background operation loop
    while (1) {
        //Use the select function to read the pipeline without blocking
        FD_ZERO(&set);
        FD_SET(fd, &set);
        timeout.tv_sec = 1;
        timeout.tv_usec = 0;
        int result = select(fd + 1, &set, NULL, NULL, &timeout);
        if (result == -1) {
            fprintf(stderr, "Failed to select pipe: %s\n", strerror(errno));
            exit(EXIT_FAILURE);
        } else if (result == 0) {
            // Pipeline has no data, continue with background operation
            printf("No data in pipe, continue.\n");
            sleep(1);
            printf("Continue other work...\n");
        } else {
            // There is data in the pipeline, output and continue with background operations
            ssize_t len = fread(buffer, 1, BUFFER_SIZE - 1, nmap_pipe);
            if (len == -1) {
                fprintf(stderr, "Failed to read pipe: %s\n", strerror(errno));
                exit(EXIT_FAILURE);
            } else if (len == 0) {
                // The pipeline has been closed
                printf("Pipe closed.\n");
                break;
            } else {
                buffer[len] = '\0';
                printf("Data in pipe: %s\n", buffer);
                sleep(1);
                printf("Continue other work...\n");
            }
        }
    }

    // Close pipeline
    pclose(nmap_pipe);

    return 0;
}
```
